haskell/src/Monocle/Api/Server.hs

```diff
@@ -28,13 +28,65 @@ import qualified Monocle.UserGroup as UserGroupPB
 import Monocle.Version (version)
 import Proto3.Suite (Enumerated (..))
 
+-- | 'getConfig' reload the config automatically from the env
+getConfig :: AppM Config.Config
+getConfig = do
+  loadConfig <- asks config
+  (reloaded, config) <- liftIO loadConfig
+  crawlerReloadStatus <- asks cRStatus
+  when reloaded $ do
+    liftIO $
+      -- All indexed must be reloaded
+      atomically
+        . writeTVar
+          crawlerReloadStatus
+        $ (\ws -> (Config.getWorkspaceName ws, True)) <$> Config.getWorkspaces config
+  pure config
+
+-- | 'updateCrawlerMD' refresh crawler Metadata if needed
+updateCrawlerMD :: Config.Index -> AppM ()
+updateCrawlerMD index = do
+  shouldReload <- getCrawlerMDNeedReload
+  when shouldReload $ do
+    refreshCrawlerMD
+    setCrawlerMDReloaded
```

Comment:
It looks like this is not enough to prevent concurrent update. I think this whole function needs to happen in a `atomically . stateTVar` so that the first thread would do the initCrawlerMetadata when the bool is False, effectively blocking the other thread, which should then be able to skip the init phase.